contracts/smart-contract-wallet/SmartAccount.sol

```diff
@@ -218,28 +222,28 @@ contract SmartAccount is
                     // Signature info
                     nonces[1]++
                 );
-            // Execute transaction.
             txHash = keccak256(txHashData);
-
             checkSignatures(txHash, signatures);
         }
 
-
         // We require some gas to emit the events (at least 2500) after the execution and some to perform code until the execution (500)
         // We also include the 1/64 in the check that is not send along with a call to counteract potential shortings because of EIP-150
-        require(gasleft() >= max((_tx.targetTxGas * 64) / 63,_tx.targetTxGas + 2500) + 500, "BSA010");
+        // Bitshift left 6 bits means multiplying by 64, just more gas efficient
+        if(gasleft() < max((_tx.targetTxGas << 6) / 63,_tx.targetTxGas + 2500) + 500) 
+            revert NotEnoughGasLeft(gasleft(), max((_tx.targetTxGas << 6) / 63,_tx.targetTxGas + 2500) + 500);
         // Use scope here to limit variable lifetime and prevent `stack too deep` errors
         {
             // If the gasPrice is 0 we assume that nearly all available gas can be used (it is always more than targetTxGas)
             // We only substract 2500 (compared to the 3000 before) to ensure that the amount passed is still higher than targetTxGas
             success = execute(_tx.to, _tx.value, _tx.data, _tx.operation, refundInfo.gasPrice == 0 ? (gasleft() - 2500) : _tx.targetTxGas);
             // If no targetTxGas and no gasPrice was set (e.g. both are 0), then the internal tx is required to be successful
             // This makes it possible to use `estimateGas` without issues, as it searches for the minimum gas where the tx doesn't revert
-            require(success || _tx.targetTxGas != 0 || refundInfo.gasPrice != 0, "BSA013");
+            if(!success && _tx.targetTxGas == 0 && refundInfo.gasPrice == 0) revert CanNotEstimateGas(_tx.targetTxGas, refundInfo.gasPrice, success);
```

Comment:
we can leave this as require if there isn't significant improvement. 
anyway taking a note so it helps with resolving conflicts

Reply:
you mean no significant gas improvement?

Reply:
hmm yes, just for execTransaction to not alter from safe internal method. we can keep here now to be consistent and later review reverts altogether.